### ISSUE TITLE
Storing reference to Python function causes memory leak

### ISSUE THREAD

**Issue by djbarker** [ninia/jep#99]
I have a Java function which accepts a `Function` as an argument. 
In Python I define a function and pass this into the Java function. 
The Java function stores the `Function` in a variable which can later be called successfully. 
Upon closing the Jep instance I can no longer call the function in Java; Jep rightfully throws an `UndeclaredThrowableException` with a message "jep.JepException: Jep instance has been closed.". However it seems that any memory used by the Jep instance for Python is not released. 

Here is some code which demonstrates what I see on my machine

```java
package jep_test;

import jep.Jep;

public class JepCircularRefIssue {

    public static interface Applicator<T> {
        public void apply(T value);
    }

    static Applicator<?> store;

    public <T> void javaFunction(Applicator<T> value) {
        store = value;
    }

    public static void main(String[] argc) throws Exception {
        Jep jep = new Jep(false);
        jep.eval("def f(x):\n\tprint('here');");
        jep.eval("from jep_test import JepCircularRefIssue");
        jep.eval("x = [1]*10_000_000");  // memory usage jumps up here by ~75 MB ...
        //jep.eval("JepCircularRefIssue().javaFunction(f)");  // (uncommenting this stops memory decreasing)
        jep.close(); // ... and decreases again here by the same amount
    }

}
```

Since we cannot call the function after calling `jep.close()` anyway, I would still expect the memory to be cleared. 

I also suspect there is a secondary issue here, in that any Java objects which are in scope for the Python function `f` also get kept around.

- OS Platform, Distribution, and Version:
Linux, Fedora 24
- Python Distribution and Version:
Python 3.6 (Anaconda)
- Java Distribution and Version:
JVM 1.8.0_66

**Comment by ndjensen** [comment]
Thanks for reporting this and providing a unit test.  @bsteffensmeier and I looked at it and found the leak.  I put it on the dev_3.7 branch to be ready for a 3.7.1 release, can you try it out?  https://github.com/ninia/jep/commit/d215c71e30b30f7340ce38a760c024eca5613c4d

I tested your test without x but with a for loop going on forever and watched the memory continually climb, so my test was not exactly the same as your test.  If you can confirm that fixes it that would be appreciated.  I can hopefully get to releasing 3.7.1 next week with this fix and a few other minor fixes.

**Comment by djbarker** [comment]
Hi Nathan, thanks for the very prompt reply. Sorry I didn't get to look at this over the weekend. 

Unfortunately it seems the issue is still present. As far as I can tell your commit should be a no-op; there is no return (or other flow-control) between the assignment and the for-loop so flipping them shouldn't matter, no?

Cheers,
Dan

**Comment by ndjensen** [comment]
How are you testing the memory?  There's no guarantee the OS will reclaim the memory once the Jep instance is closed, hence I put it in a for loop where it eventually plateaus.

The change is `Jep.close()` loops over the PyObjects to close them, and `PyObject.close()` calls `PyObject.decref()`.  `PyObject.decref()` then calls `isValid()` and `isValid()` fails because it calls `jep.isValidThread()` which fails because jep.closed was set to true.  So we need to ensure closed on the Jep instance is set to true after the PyObjects are closed.

**Comment by djbarker** [comment]
Yes indeed, I was also testing with a for-loop and seeing my memory continually climb. Please ignore the noise from me, it _does_ seem to be fixed. IntelliJ was still picking up my old .jar despite me pointing it to the new build. I manually copied over it and memory use now plateaus as you say. Thanks for the quick fix.

**Comment by ndjensen** [comment]
Fixed in 3.7.1.
